### PR TITLE
Public IP removal

### PR DESCRIPTION
**Pull request by ArielLima**
### What this PR does / why we need it:

Removes public IP from private clusters.

### Test plan for issue:
We deployed a full rp service in DEV with the current changes and created a cluster.

### CODE REVIEW

**Comment by ArielLima** [comment]
This is very much a WIP and there is no need of reviews at this time.

**Comment by ArielLima** [comment]
Cards created for this feature:
https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/15054503
https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/15058440

File: pkg/installer/custominstallconfig.go
Comment (bennerv) [inline pkg/installer/custominstallconfig.go:72]: This shouldn't go here.  This function is specifically for creating custom installation configuration, and we're not doing that with the UserDefinedRouting things.  We should be doing this elsewhere.

File: pkg/installer/deployresources_resources.go
Comment (bennerv) [inline pkg/installer/deployresources_resources.go:21]: Since we don't require a load balancer anymore, we can just gate on if we're using UserDefinedRouting, then don't add the ip configuration to point to the existing load balancer.

File: pkg/installer/generateconfig.go
Comment (bennerv) [inline pkg/installer/generateconfig.go:114]: nit - don't use upper-case variable names.  
```suggestion
	outboundType := azuretypes.LoadbalancerOutboundType

File: pkg/installer/generateconfig.go
Comment (bennerv) [inline pkg/installer/generateconfig.go:230]: Don't actually need to change this, as the case doesn't differ.

File: pkg/installer/generateconfig.go
Comment (bennerv) [inline pkg/installer/generateconfig.go:190]: what happens if outboundType is empty?

File: pkg/installer/install.go
Comment (bennerv) [inline pkg/installer/install.go:40]: +1 on removing `ctx` as it's not actually used.

**Review by bennerv — CHANGES_REQUESTED** [review]
good work so far!  Few things I noticed

File: pkg/installer/custominstallconfig.go
Comment (bennerv) [reply pkg/installer/custominstallconfig.go:72]: Make it it's own step within `pkg/cluster/install.go`.

Call it something like: `populateUserDefinedRouting` and utilize the existing `pkg/util/feature/subscription.go` function called `IsRegisteredForFeature` instead of writing your own.

Comment (ArielLima) [reply pkg/installer/custominstallconfig.go]: +1

File: pkg/installer/deployresources_resources.go
Comment (ArielLima) [reply pkg/installer/deployresources_resources.go:21]: I think we are already doing that

File: pkg/installer/generateconfig.go
Comment (ArielLima) [reply pkg/installer/generateconfig.go]: +1

File: pkg/installer/generateconfig.go
Comment (ArielLima) [reply pkg/installer/generateconfig.go]: Can't be empty, we initialize it here:
```go
// determine outbound type based on cluster visibility
outboundType := azuretypes.LoadbalancerOutboundType
if m.oc.Properties.NetworkProfile.UserDefinedRouting == api.UserDefinedRoutingDisabled {
	outboundType = azuretypes.UserDefinedRoutingOutboundType
}
```

File: pkg/api/admin/openshiftcluster_convert.go
Comment (bennerv) [inline pkg/api/admin/openshiftcluster_convert.go:196]: Need to add a conversion in `toExternal` as well so we can view it from an admin API perspective.

File: pkg/installer/deployresources_resources.go
Comment (bennerv) [inline pkg/installer/deployresources_resources.go:23]: nit:
lowercase variable names `lbBackendAddressPool`

File: pkg/installer/install.go
Comment (bennerv) [inline pkg/installer/install.go:53]: can you pull this into it's own function to make things a bit cleaner?

**Review by bennerv — CHANGES_REQUESTED** [review]
One more thing:

`pkg/api/defaults.go` will need a case if the field is empty to set the `UserDefinedRouting` to disabled.

File: pkg/installer/install.go
Comment (ArielLima) [reply pkg/installer/install.go]: +1

File: pkg/installer/deployresources_resources.go
Comment (ArielLima) [reply pkg/installer/deployresources_resources.go]: +1

File: pkg/api/admin/openshiftcluster_convert.go
Comment (ArielLima) [reply pkg/api/admin/openshiftcluster_convert.go:197]: +1

File: pkg/api/admin/openshiftcluster.go
Comment (bennerv) [inline pkg/api/admin/openshiftcluster.go:149]: After some discussion with Jerome, we decided we want to do something like this instead.  This will allow us to support more outbound types in the future if desired.  I'm not sure what to call the "default".  I believe within OCP documentation it's `LoadbalancerOutboundType` so maybe that's a good thing to set it as instead of `Default`.

```
OutboundType OutboundType
```

```const (
OutboundTypeUserDefinedRouting OutboundType = "UserDefinedRouting"
OutboundTypeDefault OutboundType = "Default"
```

**Review by bennerv — COMMENTED** [review]
Couple changes asked by Jerome on the experience.

File: pkg/api/admin/openshiftcluster.go
Comment (ArielLima) [reply pkg/api/admin/openshiftcluster.go]: +1

File: pkg/api/admin/openshiftcluster.go
Comment (SudoBrendan) [inline pkg/api/admin/openshiftcluster.go:133]: nit: a slightly more informative comment would be good

File: pkg/cluster/deploystorage.go
Comment (SudoBrendan) [inline pkg/cluster/deploystorage.go:125]: nit: this line is making me think we should have a third constant:

1. User Defined
2. (whatever it is we want to call what we do right now... public?)
3. Default (which may change over time)

This `if` makes a lot more sense if we use a constant like "public" instead of "default"

**Review by SudoBrendan — CHANGES_REQUESTED** [review]
This is looking good so far, nice work! A few nits.

File: pkg/cluster/deploystorage.go
Comment (ArielLima) [reply pkg/cluster/deploystorage.go]: +1 I don't think it makes sense to introduce a third constant at this time since we either want it user defined or default. I will rename the "default" to `OutboundTypeLoadbalancerRouting` since that will be more clear.

File: pkg/api/admin/openshiftcluster.go
Comment (ArielLima) [reply pkg/api/admin/openshiftcluster.go]: +1

File: pkg/api/admin/openshiftcluster.go
Comment (bennerv) [inline pkg/api/admin/openshiftcluster.go:139]: I think we just want to call this `Loadbalancer` not `LoadbalancerRouting`.  

Maintains consistency with https://github.com/openshift/installer/blob/master/pkg/types/azure/platform.go#L18

File: pkg/util/network/network.go
Comment (bennerv) [inline pkg/util/network/network.go:11]: nit:
```suggestion
func DetermineOutboundType(oc *api.OpenShiftCluster, sub *api.Subscription) error {

File: pkg/util/network/network.go
Comment (bennerv) [inline pkg/util/network/network.go:20]: We should move this to the installation package.  This isn't really a utility.  As well, we can define it on the manager struct and call it using `m.determineOutboundType`

File: pkg/installer/install.go
Comment (bennerv) [inline pkg/installer/install.go:50]: This has to happen before `m.applyInstallConfigCustomisations` otherwise the field will be nil.

File: pkg/util/network/network.go
Comment (bennerv) [inline pkg/util/network/network.go:1]: I think this file isn't used anywhere now and can be deleted.

**Review by bennerv — CHANGES_REQUESTED** [review]
Almost there - great work!

File: pkg/installer/install.go
Comment (bennerv) [inline pkg/installer/install.go:40]: This has to happen before we generate the installation configuration since the `generateInstallConfig` function uses the outbound type to set the installConfig outboundType parameter.  Otherwise it'll nil out and fail? or default to something bad.

File: pkg/cluster/deploystorage.go
Comment (bennerv) [inline pkg/cluster/deploystorage.go:126]: So the ordering on this is interesting.  Can you append to your comment and add something like:
`m.determineOutboundType` will ensure that the outbound type is set correctly so we do not have to evaluate whether the visibility is public or private in this if statement since it's already been confirmed by the function.`

It's very order-specific and it's confusing to review/read in isolation.  We have to look at the installation as a whole to ensure this is doing what we expect.

**Review by bennerv — CHANGES_REQUESTED** [review]
Super close.  Just noticed something on it that won't work.

**Comment by cblecker** [comment]
Would you mind rebasing, @ArielLima? Removing ready-to-review until that's done.

**Comment by ArielLima** [comment]
@cblecker As far as I am aware this is on hold until hive is finished so maybe wait until then to rebase?

File: pkg/api/admin/openshiftcluster.go
Comment (bennerv) [inline pkg/api/admin/openshiftcluster.go:150]: can you make this mutable so we can modify it on PUCM for existing clusters.  Then we can set it for the clusters we performed this action on manually.  


In the following release (after this one is merged), we will remove the field.

**Review by bennerv — COMMENTED** [review]
Can you open up a PR to https://github.com/hawkowl/ARO-Installer as well to pull in the installer related chagnes (pkg/installer)?  And then just add the mutable flag too then LGTM - ship it.

File: pkg/api/openshiftcluster.go
Comment (bennerv) [inline pkg/api/openshiftcluster.go:267]:
```suggestion
	OutboundType           OutboundType           `json:"outboundType,omitempty"`

File: pkg/cluster/networkprofile.go
Comment (bennerv) [inline pkg/cluster/networkprofile.go:71]: this line is redundant since you're already doing this on line 75 below.

**Review by bennerv — CHANGES_REQUESTED** [review]
@ArielLima  can we get these changes and the other comments fixed so we can merge this?

**Comment by bennerv** [comment]
Also, please update the PR description and link to appropriate work items.

File: pkg/api/openshiftcluster.go
Comment (bennerv) [inline pkg/api/openshiftcluster.go:267]:
```suggestion
	OutboundType           OutboundType           `json:"outboundType,omitempty"`

**Review by bennerv — COMMENTED** [review]
Just one capitalization issue and I'm good to approve.  Great work here.

**Review by bennerv — APPROVED** [review]
lgtm

File: pkg/cluster/deploystorage.go
Comment (s-amann) [inline pkg/cluster/deploystorage.go:140]: why do we name this one with the `-pip-v4` suffix? does it have specific meaning?

**Review by s-amann — COMMENTED** [review]
just some questions - forgot to submit them before it got merged..

File: pkg/installer/custominstallconfig.go
Comment (bennerv) [reply pkg/installer/custominstallconfig.go:23]: Context isn't used anywhere within the function.  Just asked to remove it.  Could've been a separate commit, but left it in there.

File: pkg/cluster/deploystorage.go
Comment (bennerv) [reply pkg/cluster/deploystorage.go:146]: https://msazure.visualstudio.com/AzureRedHatOpenShift/_wiki/wikis/ARO.wiki/375424/Remove-public-IP-manually#

Comment (bennerv) [reply pkg/cluster/deploystorage.go:146]: >I believe a public cluster would end up with both -pip-v4 and -default-v4.

This is the desired behavior. Existing public ingress & apiserver clusters have both these IP addresses.  We don't want to change the default behavior we've been using for existing clusters.  The default IP address is for ingress, the other is used for outbound explicit rule.

File: pkg/cluster/deploystorage.go
Comment (bennerv) [reply pkg/cluster/deploystorage.go:140]: It's how upstream OCP installed named them.  It doesn't have a specific meaning.

**Pull request merged**
